src/TabComplete.js

```diff
@@ -0,0 +1,306 @@
+        // actually have things to tab over
+        return this.completing && this.matchedList.length > 1;
+    }
+
+    stopTabCompleting() {
+        this.completing = false;
+        this.currentIndex = 0;
+        this._notifyStateChange();
+    }
+
+    startTabCompleting() {
+        this.completing = true;
+        this.currentIndex = 0;
+        this._calculateCompletions();
+    }
+
+    /**
+     * Do an auto-complete with the given word. This terminates the tab-complete.
+     * @param {string} someVal
+     */
+    completeTo(someVal) {
+        this.textArea.value = this._replaceWith(someVal, true);
+        this.stopTabCompleting();
+        // keep focus on the text area
+        this.textArea.focus();
+    }
+
+    /**
+     * @param {Number} numAheadToPeek Return *up to* this many elements.
+     * @return {Entry[]}
+     */
+    peek(numAheadToPeek) {
+        if (this.matchedList.length === 0) {
+            return [];
+        }
+        var peekList = [];
+
+        // return the current match item and then one with an index higher, and
+        // so on until we've reached the requested limit. If we hit the end of
+        // the list of options we're done.
+        for (var i = 0; i < numAheadToPeek; i++) {
+            var nextIndex;
+            if (this.opts.allowLooping) {
+                nextIndex = (this.currentIndex + i) % this.matchedList.length;
+            }
+            else {
+                nextIndex = this.currentIndex + i;
+                if (nextIndex === this.matchedList.length) {
+                    break;
+                }
+            }
+            peekList.push(this.matchedList[nextIndex]);
+        }
+        // console.log("Peek list(%s): %s", numAheadToPeek, JSON.stringify(peekList));
+        return peekList;
+    }
+
+    /**
+     * @param {DOMEvent} e
+     * @return {Boolean} True if the tab complete state changed as a result of
+     * this event.
+     */
+    onKeyDown(ev) {
+        var wasInPassiveMode = this.inPassiveMode && !ev.passive;
+        this.inPassiveMode = ev.passive;
+
+        if (ev.keyCode !== KEY_TAB) {
+            if (this.completing && ev.keyCode !== KEY_SHIFT &&
+                    !ev.metaKey && !ev.ctrlKey && !ev.altKey && ev.keyCode !== KEY_WINDOWS) {
+                // they're resuming typing; reset tab complete state vars.
+                this.stopTabCompleting();
+
+                // fall through to auto-enter-tab-completing if set
+                if (!this.opts.autoEnterTabComplete) {
+                    return true;
+                }
+            }
+
+            if (this.opts.autoEnterTabComplete) {
+                clearTimeout(this.enterTabCompleteTimerId);
+                this.enterTabCompleteTimerId = setTimeout(() => {
+                    if (!this.completing) {
+                        // inject a fake tab event so we use the same code paths
+                        // as much as possible. Add a 'passive' flag to indicate
```

Comment:
Yeah I'm not really sure what to do with this. I like reusing the code path because there *is* stuff which always needs to be done (sanity checking the `textArea`, `startTabCompletion`, `_notifyStateChange`, etc).

I refute the statement "so much special-casing" given this flag results in precisely 2 changes to a normal `onKeyDown` event (setting the internal state var `inPassiveMode` and then conditionally changing the `nextMatchedEntry` input arg). I do think this has a risk of bloating over time to be not-so-trivial, but equally there is the risk that changes to how tab-complete is being handled will not be applied to both passive tabs and active tabs.